src/packages/@ncigdc/utils/auth/awg.js

```diff
@@ -0,0 +1,137 @@
+  localStorage.setItem('awg_sessionLastRenewed', now);
+  return !!localStorage.awg_sessionLastRenewed;
+};
+
+const removeSessionRenewalTimestamp = () => {
+  localStorage.removeItem('awg_sessionLastRenewed');
+  return !localStorage.awg_sessionLastRenewed;
+};
+
+const resetSessionRenewalTimestamp = now => (
+  removeSessionRenewalTimestamp() &&
+  createSessionRenewalTimestamp(now)
+);
+
+/* this is meant to avoid duplicate queries, because having multiple Relay Environments
+ * means renewal race conditions not worth dealing with at this point. ¯\_(ツ)_/¯
+ */
+let queryCache = null;
+const handleQueryCache = value => {
+  // hold and use the latest response for 5 seconds;
+  queryCache = value;
+
+  setTimeout(() => {
+    queryCache = null;
+  }, 5000);
+
+  return queryCache;
+};
+
+const renewAWGToken = throttle(({
+  now,
+  origin,
+  timeSinceLastRenewal,
+}) => {
+  consoleDebug(`Querying AWG fence at ${now}, by ${origin}`);
+
+  return queryCache || fetch(urlJoin(FENCE, 'user'), {
+    credentials: 'include',
+  })
+    .then(response => {
+      const responseTimestamp = moment.utc().format();
+      consoleDebug(`response on AWG token renewal at ${responseTimestamp}, triggered by ${origin}`);
+      consoleDebug(response);
+
+      response.ok
+        ? resetSessionRenewalTimestamp(responseTimestamp)
+        : removeSessionRenewalTimestamp();
+
+      return handleQueryCache(response.ok);
+    })
+    .catch(err => {
+      console.error(`Error ${
+        timeSinceLastRenewal ? 'renewing' : 'creating'
+      } the AWG token`, err);
+      timeSinceLastRenewal && console.error(`Time since last renewal: ${
+        timeSinceLastRenewal / 1000
+      } seconds`);
+      removeSessionRenewalTimestamp();
+
+      throw handleQueryCache(err);
+    });
+}, 1000);
+
+export const checkAWGSession = (renewalType = 'manual', origin = 'unknown origin') => {
+  if (localStorage.awg_sessionLastRenewed) {
+    const nowMoment = moment.utc();
+    const now = nowMoment.format();
+    const timeSinceLastRenewal = nowMoment.diff(moment(
+      localStorage.awg_sessionLastRenewed.replace('-pending', ''), // '-pending' throttles queries
+    ));
+
+    consoleDebug(`Checking AWG token at ${now}, ${renewalType}. Time since last renewal is ${timeSinceLastRenewal}`);
+    consoleDebug(`Origin: ${origin}`);
+
+    return timeSinceLastRenewal < AWG_TOKEN_EXPIRY
+        ? (
+          consoleDebug(`Token set less than ${
+            AWG_TOKEN_EXPIRY / 60 / 1000
+          } minutes ago, ${timeSinceLastRenewal / 1000} seconds`),
+          'Token shouldn\'t need renewal yet'
+        )
+        : (
+          consoleDebug(`Attempting token renewal at ${now}, by ${origin}`),
+          renewAWGToken({
+            now,
+            origin,
+            timeSinceLastRenewal,
+          })
+        );
+  }
+  return !AWG; // do not interfere if not on AWG
+};
+
+export const clearAWGSession = timerId => {
```

Comment:
I don't think anything other than the `RelaySetup` component knows the timer ID to pass it into `clearAWGSession`. If something else calls this before `RelaySetup` is unmounted, it will clear `localStorage.awg_sessionLastRenewed`, and then this won't clear the timer when `RelaySetup` unmounts. It might be better to keep the timer management in `Root.js`.

This is not a blocking issue since we shouldn't be enabling the refresh timer in any actual deployments anyway.

Reply:
There's no timer in normal runtime. the timer is only even created if we're in testing/debugging mode.
This is functionality I've added to help me work. Not functional in prod (unless enabled)